### PR TITLE
Blockchain project

### PR DESCRIPTION
**Pull request by glouppe**

### CODE REVIEW

File: project/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+### Group formation
+
+Please register your group at the [Montefiore submission system](https://submit.montefiore.ulg.ac.be/), also if you are working alone.
+
+---
+
+## Evaluation
+
+This project is a requirement to pass INFO8002, and **must** be completed and submitted with all deliverables to the [Montefiore submission system](https://submit.montefiore.ulg.ac.be/) by the **hard deadline** of `21/12/2018 23:59`. You are allowed to implement the project by yourself, or in groups with a maximum of *3* students (see [group formation](#group-formation)).
```
Comment (glouppe) [inline project/README.md]: Merge 'Group formation' and 'Evaluation'? (and put everything under Instructions)

File: project/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+We expect the following deliverables;
+
+- An **implementation** with the following requirements:
+  * a Python implementation,
+  * Blockchain and Block architecture with accompanying consensus rules,
+  * a bootstrapping procedure for new nodes joining the network,
+    - New nodes have to "download" the Blockchain from peers.
+  * proof of work (PoW) with a parameterizable difficulty level,
+  * abstraction layer for the top-level application:
+    - `put(key, value)`
```
Comment (glouppe) [inline project/README.md]: Specify the versioning semantics?

File: project/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    Stores the `value` with the associated `key` in the Blockchain. Please make note of the fact that the `put` operation doesn't necessarily deliveres after it has been completed. It may therefore be a good idea to add a `callback` procedure (or a different mechanism) that is called whenever the key has been added to the chain, or if a failure occurred (i.e., duplicate key in the same block). This serves the purpose of the `deliver` operation from the lectures.
+    - `get(key)`
+    Searches the Blockchain for the latest value with the specified key.
+    - `get_all(key)`
+    Retrieves all values from the Blockchain with the specified key.
+  * code the reproduce the experiments and the associated plots,
+  * broadcast implementation to dissipate information through your Blockchain network (choose wisely),
```
Comment (glouppe) [inline project/README.md]: dissipate -> disseminate

File: project/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+- A **report** in PDF format that must include:
+  * a detailed description of the individual components in your solution,
+  * a discussion of the consequences and scalability of your broadcast implementation,
+  * a discussion about the applicability of Blockchain to this problem,
+  * an experiment showning the transaction throughput with respect to the difficulty level,
+  * an experiment demonstrating the resistance against Bryzantine attacks under different difficulty levels (e.g., double spending with 51% attack).
+
```
Comment (glouppe) [inline project/README.md]: Nowhere it is told they have to implement the given REST API. This should be clarified.

File: project/code/keychain/blockchain.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+import hashlib
+import multilib
+import numpy as np
+
+from keychain.util import hash
+
+
+
+class Block:
```
Comment (glouppe) [inline project/code/keychain/blockchain.py:8]: Is this given? or shall we rather just specify the class with empty methods they should fill in? (in which case, we should document exactly what each method does)

File: project/code/keychain/blockchain.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        return self._data[key]
+
+    def __getitem__(self, key):
+        return self.value(key)
+
+    def is_valid(self):
+        raise NotImplementedError
+
+
+class Transaction:
```
Comment (glouppe) [inline project/code/keychain/blockchain.py:22]: same

File: project/code/keychain/blockchain.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+    def __init__(self, key, value):
+        self.key = key
+        self.value = value
+
+    def hash(self):
+        return hash(self)
+
+
+class Blockchain:
```
Comment (glouppe) [inline project/code/keychain/blockchain.py]: same

**Review by glouppe — COMMENTED** [review]
Where is the REST related stuff?

**Comment by glouppe** [comment]
CC: @JoeriHermans 

Since we want them to be able think for themselves and be able to design systems, what would you think of giving them minimal instructions? In particular, no code provided -- we know now that this is doable.

We could broadly ask them to implement a simple distributed versioned key-value store based on a blockchain. 
- The system should be designed around 3 modules they have to implement (broadcast, blockchain and the high-level key-value store). 
- The consensus by proof-of-work should have a parameterizable level of difficulty.
- The modules must be implemented behind a REST API. 
- Instructions (i.e. scripts) should be provided (by them) to launch, bootstrap and use the key-value store.
- They are otherwise free to design and implement things in any way they see fit, but Keep It Stupid Simple.

From this, we would ask them to:
- demonstrate the effect of the difficulty in terms of computations and delay (must provide scripts, plots and a critical discussion).
- demonstrate the resilience of the key-value store to faults (must provide scripts, plots and a critical discussion).
- simulate a 51% attack (must provide scripts, plots and a discussion).

What do you think? (If you agree, then disregard my previous comments above regarding your commits)

In a tutorial session, you could show them how to implement a REST API using e.g. Flask.

**Comment by glouppe** [comment]
Or is it too much?

File: project/README.md (hunk 4 above)
Comment (JoeriHermans) [reply project/README.md]: Clarified, implementing API.

File: project/code/keychain/blockchain.py (hunk 5 above)
Comment (JoeriHermans) [reply project/code/keychain/blockchain.py:8]: At the moment this is "given", but maybe this makes it a bit trivial.

File: project/README.md (hunk 4 above)
Comment (JoeriHermans) [reply project/README.md]: API implemented.

**Comment by JoeriHermans** [comment]
@glouppe Proposition is good, but maybe they could use the code I have now as some form of boilerplate code which they can change.

I'll also give them a tutorial how to use Flask and the `requests` module if required.

**Comment by JoeriHermans** [comment]
@glouppe Could you review the README?

**Comment by glouppe** [comment]
We were editing the README in parallel :) I made some more updates. Can you check it? 

For the code, I would provide them with a minimal stub and instead let them design the system as they wish. The existence of this stub should be specified in the instructions.

**Comment by glouppe** [comment]
> For the code, I would provide them with a minimal stub and instead let them design the system as they wish. The existence of this stub should be specified in the instructions.

Therefore, can you trim the code as much as possible? 

Also, provide detailed docstrings for what is left (in particular, the `Storage` API).

**Comment by JoeriHermans** [comment]
@glouppe Oh sorry. Added a few other things I missed in my original version. Stripping the code now, will ping you when it is ready.

**Comment by JoeriHermans** [comment]
@glouppe Stubs are ok for you in it's current form?

**Comment by glouppe** [comment]
I'll start reading this now, making some small changes along the way.

File: project/code/application.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    # Now the key should be available,
+    # unless a different node `put` a new value.
+    print(storage.retrieve(key))
+
+    # Show all values of the key.
+    print(storage.retrieve_all(key))
+
+    # XXX: not defined!
+    application.terminate()
```
Comment (glouppe) [inline project/code/application.py]: this is not defined. shall we remove it?

File: project/code/setup.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+"""```KeyChain``` setup file."""
```
Comment (glouppe) [inline project/code/setup.py]: is this file needed at all?

File: project/code/application.py (hunk 8 above)
Comment (JoeriHermans) [reply project/code/application.py]: Yes, forgot to remove this. Thanks.

**Comment by glouppe** [comment]
I added some comments and changes. 

I wonder if we should either:
- add a stub for the broadcast?
- add a stub of REST api?
- OR nothing like this at all, keep things like this? (i prefer this option)

**Comment by JoeriHermans** [comment]
> I added some comments and changes.
> 
> I wonder if we should either:
> 
> * add a stub for the broadcast?
> * add a stub of REST api?
> * OR nothing like this at all, keep things like this? (i prefer this option)

Preferably the last one, because the way they interact with the REST API (e.g., the mining) will depend on how they initialize the mining procedure and integrate this with the Blockchain class.

File: project/code/setup.py (hunk 9 above)
Comment (JoeriHermans) [reply project/code/setup.py]: Depends if they want to create a module or not. But typically I find it easier to do it like this, especially if you have to create experiments. Then you can just instantiate the `Storage`, and use that in your experiments.

**Comment by glouppe** [comment]
Will you be around tomorrow? Would be good to sit together for the final edits and for the publication of the project. 

(Also, I am not sure why github says there is a conflict in this branch?)

**Comment by JoeriHermans** [comment]
Yes, tomorrow is the exercise session according to the schedule.

There do not seem to be any conflicts here?

**Comment by glouppe** [comment]
Yes, tomorrow is scheduled for the exercise. We will merge by then.

File: project/code/keychain/store.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+class Storage:
+    def __init__(self, bootstrap, miner, difficulty):
+        """Allocate the backend storage of the high level API, i.e.,
+        your blockchain. Depending whether or not the miner flag has
+        been specified, you should allocate the mining process.
+        """
+        self._blockchain = Blockchain()
+        raise NotImplementedError
+
+    def put(self, key, value, block=True):
```
Comment (glouppe) [inline project/code/keychain/store.py:34]: Provided we ask them to run the blockchain behind a rest service, shouldn't we exemplify here how to make a (REST) request instead? I foresee confusion otherwise.

File: project/code/keychain/store.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        """Polls the blockchain to check if the data is available."""
+        raise NotImplementedError
+
+
+class Storage:
+    def __init__(self, bootstrap, miner, difficulty):
+        """Allocate the backend storage of the high level API, i.e.,
+        your blockchain. Depending whether or not the miner flag has
+        been specified, you should allocate the mining process.
+        """
```
Comment (glouppe) [inline project/code/keychain/store.py:30]: Shouldn't we instead say that the Storage connects to an (existing and persistent) chain, rather than allocate a new one?

File: project/code/keychain/store.py (hunk 10 above)
Comment (JoeriHermans) [reply project/code/keychain/store.py:34]: I would abstract it away within the Blockchain class, makes the API a bit cleaner imho, e.g.,

```python
def put(key, value):
    transaction = Transaction(key, value)
    self._blockchain.do_transaction(transaction)
    # `do_transaction` does the broadcast (and the appending to the block if the mining process has been specified).
```

File: project/code/keychain/store.py (hunk 11 above)
Comment (JoeriHermans) [reply project/code/keychain/store.py:30]: > Shouldn't we instead say that the Storage connects to an (existing and persistent) chain, rather than allocate a new one?

That's what the bootstrap address is for. It will download the peers from the bootstrap node, and start the downloading procedure (by calling /blockchain or something), and continue from there.

File: project/code/keychain/store.py (hunk 10 above)
Comment (glouppe) [reply project/code/keychain/store.py:34]: Alright, I rephrased the instructions to simply say that network communication should be implemented through a rest api.

File: project/code/keychain/store.py (hunk 11 above)
Comment (glouppe) [reply project/code/keychain/store.py:30]: Alright, maybe specify the purpose of `bootstrap` in the docstring?
